File: mrb_doc/audio/music.rb
# Updates the buffer for the specified music, you should call this every frame.
# @param music [Music]
# @return [nil]
def update_music_stream(music)
  # mrb_update_music_stream
  # src/mruby_integration/audio/music.cpp
  nil
end

# Pauses a music object so it can be resumed.
# @param music [Music]
# @return [nil]
def pause_music_stream(music)
  # mrb_pause_music_stream
  # src/mruby_integration/audio/music.cpp
  nil
end

# Resumes a paused music object.
# @param music [Music]
# @return [nil]
def resume_music_stream(music)
  # mrb_resume_music_stream
  # src/mruby_integration/audio/music.cpp
  nil
end

# Sets volume for a music object.
# @param music [Music]
# @param volume [Float] A value between 0.0 and 1.0.
# @return [nil]
def set_music_volume(music, volume)
  # mrb_set_music_volume
  # src/mruby_integration/audio/music.cpp
  nil
end

# Sets pitch for a music object.
# @param music [Music]
# @param pitch [Float]
# @return [nil]
def set_music_pitch(music, pitch)
  # mrb_set_music_pitch
  # src/mruby_integration/audio/music.cpp
  nil
end

# Gets the length of a music object in seconds.
# @param music [Music]
# @return [Float] The length of the music object in seconds.
def get_music_time_length(music)
  # mrb_get_music_time_length
  # src/mruby_integration/audio/music.cpp
  nil
end

# Gets the time played of a music object in seconds.
# @param music [Music]
# @return [Float] The length of the music object in seconds
def get_music_time_played(music)
  # mrb_get_music_time_played
  # src/mruby_integration/audio/music.cpp
  nil
end

File: src/mruby_integration/audio/music.cpp
#include "mruby.h"
#include "raylib.h"
#include <cstdlib>

#include "mruby_integration/models/music.hpp"
#include "mruby_integration/struct_types.hpp"

auto
mrb_update_music_stream(mrb_state* mrb, mrb_value) -> mrb_value
{
  Music* music;
  mrb_get_args(mrb, "d", &music, &Music_type);

  UpdateMusicStream(*music);

  return mrb_nil_value();
}

auto
mrb_pause_music_stream(mrb_state* mrb, mrb_value) -> mrb_value
{
  Music* music;
  mrb_get_args(mrb, "d", &music, &Music_type);

  PauseMusicStream(*music);

  return mrb_nil_value();
}

auto
mrb_resume_music_stream(mrb_state* mrb, mrb_value) -> mrb_value
{
  Music* music;
  mrb_get_args(mrb, "d", &music, &Music_type);

  ResumeMusicStream(*music);

  return mrb_nil_value();
}

auto
mrb_set_music_volume(mrb_state* mrb, mrb_value) -> mrb_value
{
  Music* music;
  mrb_float volume;
  mrb_get_args(mrb, "df", &music, &Music_type, &volume);

  SetMusicVolume(*music, volume);

  return mrb_nil_value();
}

auto
mrb_set_music_pitch(mrb_state* mrb, mrb_value) -> mrb_value
{
  Music* music;
  mrb_float pitch;
  mrb_get_args(mrb, "df", &music, &Music_type, &pitch);

  SetMusicPitch(*music, pitch);

  return mrb_nil_value();
}

auto
mrb_get_music_time_length(mrb_state* mrb, mrb_value) -> mrb_value
{
  Music* music;
  mrb_get_args(mrb, "d", &music, &Music_type);

  return mrb_float_value(mrb, GetMusicTimeLength(*music));
}

auto
mrb_get_music_time_played(mrb_state* mrb, mrb_value) -> mrb_value
{
  Music* music;
  mrb_get_args(mrb, "d", &music, &Music_type);

  return mrb_float_value(mrb, GetMusicTimePlayed(*music));
}

void
append_audio_music(mrb_state* mrb)
{
  mrb_define_method(mrb,
                    mrb->kernel_module,
                    "update_music_stream",
                    mrb_update_music_stream,
                    MRB_ARGS_REQ(1));
  mrb_define_method(mrb,
                    mrb->kernel_module,
                    "pause_music_stream",
                    mrb_pause_music_stream,
                    MRB_ARGS_REQ(1));
  mrb_define_method(mrb,
                    mrb->kernel_module,
                    "resume_music_stream",
                    mrb_resume_music_stream,
                    MRB_ARGS_REQ(1));
  mrb_define_method(mrb,
                    mrb->kernel_module,
                    "set_music_volume",
                    mrb_set_music_volume,
                    MRB_ARGS_REQ(2));
  mrb_define_method(mrb,
                    mrb->kernel_module,
                    "set_music_pitch",
                    mrb_set_music_pitch,
                    MRB_ARGS_REQ(2));
  mrb_define_method(mrb,
                    mrb->kernel_module,
                    "get_music_time_length",
                    mrb_get_music_time_length,
                    MRB_ARGS_REQ(1));
  mrb_define_method(mrb,
                    mrb->kernel_module,
                    "get_music_time_played",
                    mrb_get_music_time_played,
                    MRB_ARGS_REQ(1));
}
